crates/engine/tree/src/tree/mod.rs

```diff
@@ -1207,8 +1207,19 @@ where
                 match request {
                     EngineApiRequest::InsertExecutedBlock(block) => {
                         debug!(target: "engine::tree", block=?block.block().num_hash(), "inserting already executed block");
-                        self.state.tree_state.insert_executed(block);
+                        self.state.tree_state.insert_executed(block.clone());
                         self.metrics.engine.inserted_already_executed_blocks.increment(1);
+
+                        let block_ref = block.block();
+                        if self.state.tree_state.is_canonical(block_ref.parent_hash) {
```

Comment:
we can skip this check here because the assumption here is that the inserted block is canonical